File: src/main/java/pl/laskowski/marcin/model/dividers/ComplexDivider.java
package pl.laskowski.marcin.model.dividers;

import pl.laskowski.marcin.model.Region;
import pl.laskowski.marcin.model.Sudoku;

import java.util.HashSet;
import java.util.Set;

/**
 * Created by Marcin Laskowski.
 */

public class ComplexDivider implements SudokuDivider {

    private final SudokuDivider[] dividers;

    public ComplexDivider(SudokuDivider... dividers) {
        this.dividers = dividers;
    }

    @Override
    public Set<Region> divideIntoRegions(Sudoku sudoku) {
        Set<Region> regions = new HashSet<>();
        for (SudokuDivider divider : dividers) {
            regions.addAll(divider.divideIntoRegions(sudoku));
        }
        return regions;
    }

}

File: app/src/main/java/pl/laskowski/marcin/model/dividers/ComplexDivider.kt
package pl.laskowski.marcin.model.dividers

import pl.laskowski.marcin.model.Board
import pl.laskowski.marcin.model.Region
import java.util.HashSet

class ComplexDivider(
    private vararg val dividers: SudokuDivider
) : SudokuDivider {

    override fun divideIntoRegions(board: Board): Set<Region> {
        val regions: MutableSet<Region> = HashSet()
        for (divider in dividers) {
            regions.addAll(divider.divideIntoRegions(board))
        }
        return regions
    }
}